### PR TITLE
Elasticache facts module

### LINKED ISSUE #30373: AWS ElastiCache elasticache_facts module
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
- AWS elasticache_facts module

##### SUMMARY
- Currently we don't have any way to query facts from AWS ElastiCache like we have for efs_facts, ec2_facts, etc.

### PR DESCRIPTION
**Pull request by willthames**
##### SUMMARY
New Elasticache facts module

Fixes #30373

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
elasticache_facts

##### ANSIBLE VERSION
```
ansible 2.5.0 (devel 22f3898412) last updated 2017/09/18 09:26:57 (GMT +1000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/will/src/ansible/lib/ansible
  executable location = /home/will/src/ansible/bin/ansible
  python version = 2.7.13 (default, Sep  5 2017, 08:53:59) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]

```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/elasticache_facts.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+    account_id = get_aws_account_id(module)
+    results = []
+    for cluster in clusters:
+
+        cluster = camel_dict_to_snake_dict(cluster)
+        arn = "arn:aws:elasticache:%s:%s:cluster:%s" % (region, account_id, cluster['cache_cluster_id'])
+        try:
+            tags = get_elasticache_tags_with_backoff(client, arn)
+        except botocore.exceptions.ClientError as e:
+            module.fail_json(msg="Couldn't get tags for cluster %s: %s" % (cluster['cache_cluster_id'], e.message),
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/elasticache_facts.py]: Using e.message causes exception handling failure on python3.
I deleted my cluster mid-call to get an example of this:
```
The full traceback is:
Traceback (most recent call last):
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_4fesqkxe/ansible_module_elasticache_facts.py", line 286, in get_elasticache_clusters
    tags = get_elasticache_tags_with_backoff(client, arn)
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_4fesqkxe/ansible_modlib.zip/ansible/module_utils/cloud.py", line 154, in retry_func
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_4fesqkxe/ansible_module_elasticache_facts.py", line 252, in get_elasticache_tags_with_backoff
    return client.list_tags_for_resource(ResourceName=cluster_id)['TagList']
  File "/Users/shertel/Workspace/ansible/venv/python3.5.2/lib/python3.5/site-packages/botocore/client.py", line 312, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/Users/shertel/Workspace/ansible/venv/python3.5.2/lib/python3.5/site-packages/botocore/client.py", line 601, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.errorfactory.CacheClusterNotFoundFault: An error occurred (CacheClusterNotFound) when calling the ListTagsForResource operation: shertel-ec-0001-001 is either not present or not available

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_4fesqkxe/ansible_module_elasticache_facts.py", line 319, in <module>
    main()
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_4fesqkxe/ansible_module_elasticache_facts.py", line 315, in main
    module.exit_json(elasticache_clusters=get_elasticache_clusters(client, module, region))
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_4fesqkxe/ansible_module_elasticache_facts.py", line 288, in get_elasticache_clusters
    module.fail_json(msg="Couldn't get tags for cluster %s: %s" % (cluster['cache_cluster_id'], e.message),
AttributeError: 'CacheClusterNotFoundFault' object has no attribute 'message'
```
Another reason to use AnsibleAWSModule and move toward [smart exception handling](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/aws/core.py#L124-L127).

File: lib/ansible/modules/cloud/amazon/elasticache_facts.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+    )
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True)
+
+    if not HAS_BOTO3:
+        module.fail_json(msg='boto3 and botocore are required.')
+    try:
+        region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
+        client = boto3_conn(module, conn_type='client', resource='elasticache',
+                            region=region, endpoint=ec2_url, **aws_connect_kwargs)
+    except botocore.exceptions.NoCredentialsError as e:
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/elasticache_facts.py]: `except (botocore.exceptions.NoCredentialsError, botocore.exceptions.ProfileNotFound) as e:`

File: lib/ansible/modules/cloud/amazon/elasticache_facts.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+            name=dict(required=False),
+        )
+    )
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True)
+
+    if not HAS_BOTO3:
+        module.fail_json(msg='boto3 and botocore are required.')
+    try:
+        region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
+        client = boto3_conn(module, conn_type='client', resource='elasticache',
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/elasticache_facts.py]: Should probably check if region exists before this line.

**Review by s-hertel — CHANGES_REQUESTED** [review]

**Comment by willthames** [comment]
Thanks for the review @s-hertel - now that Ansible 2.4 has been released, it makes a lot of sense to migrate to AnsibleAWSModule - I'll update the PR and let you know when it's done.

File: lib/ansible/modules/cloud/amazon/elasticache_facts.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(
+        dict(
+            name=dict(required=False),
+        )
+    )
+    module = AnsibleAWSModule(argument_spec=argument_spec, supports_check_mode=True)
+
+    if not HAS_BOTO3:
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/elasticache_facts.py]: You can remove this (and the import) now that this uses AnsibleAWSModule.

**Review by s-hertel — APPROVED** [review]

**Comment by JulianGindi** [comment]
What is the current status of this? We are interested in using this functionality and can help out with its development if needed.

**Comment by willthames** [comment]
@JulianGindi from my point of view it's complete. I suggest using it (you can just copy it into your library path) and providing feedback here (positive or negative) before merge (and after merge, add a new issue or PR)

**Review by ryansb — APPROVED** [review]

**Comment by ryansb** [comment]
Merged as 921cc6d650

**Comment by JulianGindi** [comment]
My only concern with this so far is that it returns data in a different structure than the `elasticache` module.

For example, if I want to get a cache_nodes endpoint address I would do this with the elasticache module (the top level elasticache variable is the result of a cache instance creation):

`elasticache.elasticache.data.CacheNodes[0].Endpoint.Address`

with the elasticache_facts module, it is different:

`elasticache.elasticache.elasticache_clusters.cache_nodes[0].endpoint.address`

This inconsistency makes it difficult to do a "lookup" and use the resulting data if it does exist, or create a new instance if not.

**Pull request closed without merging**